--- nginx.conf
+++ nginx.conf.mod
@@ -245,7 +245,7 @@
 
         location ~ ^/service/(?<serviceid>[0-9a-zA-Z-.]+)$ {
             # Append slash and 301-redirect.
-            rewrite ^/service/(.*)$ /service/$1/ permanent;
+            rewrite ^/service/(.*)$ $scheme://$http_host/service/$1/ permanent;
         }
 
         location ~ ^/service/(?<serviceid>[0-9a-zA-Z-.]+)/(?<url>.*) {
@@ -273,8 +273,8 @@
             proxy_set_header        X-Forwarded-Proto $scheme;
 
             proxy_pass $serviceurl;
-            proxy_redirect $servicescheme://$host/service/$serviceid/ /service/$serviceid/;
-            proxy_redirect $servicescheme://$host/ /service/$serviceid/;
+            proxy_redirect $servicescheme://$host/service/$serviceid/ $scheme://$http_host/service/$serviceid/;
+            proxy_redirect $servicescheme://$host/ $scheme://$http_host/service/$serviceid/;
             proxy_redirect / /service/$serviceid/;
 
             # Disable buffering to allow real-time protocols
